drivers/include/hmc5883l.h

```diff
@@ -0,0 +1,315 @@
+    HMC5883L_OP_MODE_IDLE       = 0x02, /**< Idle mode */
+
+} hmc5883l_op_mode_t;
+
+/**
+ * @brief   Gain (determine the sensitivity and the range)
+ * Values correspond to bits <7:5> of HMC5883L_REG_CFG_B_GN register
+ */
+typedef enum {
+    HMC5883L_GAIN_1370 = 0x00,  /**< Range +-0.88 Gs, Resolution 0.73 mG/LSb */
+    HMC5883L_GAIN_1090 = 0x20,  /**< Range  +-1.3 Gs, Resolution 0.92 mG/LSb */
+    HMC5883L_GAIN_820  = 0x40,  /**< Range  +-1.9 Gs, Resolution 1.22 mG/LSb */
+    HMC5883L_GAIN_660  = 0x60,  /**< Range  +-2.5 Gs, Resolution 1.52 mG/LSb */
+    HMC5883L_GAIN_440  = 0x80,  /**< Range  +-4.0 Gs, Resolution 2.27 mG/LSb */
+    HMC5883L_GAIN_390  = 0xa0,  /**< Range  +-4.7 Gs, Resolution 2.56 mG/LSb */
+    HMC5883L_GAIN_330  = 0xc0,  /**< Range  +-5.6 Gs, Resolution 3.03 mG/LSb */
+    HMC5883L_GAIN_230  = 0xe0,  /**< Range  +-8.1 Gs, Resolution 4.35 mG/LSb */
+} hmc5883l_gain_t;
+
+/**
+ * @brief   Magnetic field values in micro-Gauss (uGs)
+ */
+typedef struct {
+
+    int32_t x;  /**< magnetic field x-axis */
+    int32_t y;  /**< magnetic field y-axis */
+    int32_t z;  /**< magnetic field y-axis */
+
+} hmc5883l_data_t;
+
+/**
+ * @brief   Raw data set as two complements
+ */
+typedef struct {
+
+    int16_t x; /**< magnetic field x-axis as 16 bit two's complements */
+    int16_t y; /**< magnetic field y-axis as 16 bit two's complements */
+    int16_t z; /**< magnetic field z-axis as 16 bit two's complements */
+
+} hmc5883l_raw_data_t;
+
+#if MODULE_HMC5883L_INT || DOXYGEN
+
+/**
+ * @brief   HMC5883L DRDY interrupt callback function type
+ *
+ * Function prototype for the function which is called on DRDY interrupt if
+ * the interrupt is activated by #hmc5883l_init_int and the interupt pin is
+ * defined in device initialization parameters.
+ *
+ * @note The @p cb function is called in interrupt context. The application
+ *       should do nothing time consuming and not directly access sensor data.
+ */
+typedef void (*hmc5883l_drdy_int_cb_t)(void *);
+
+#endif /* MODULE_HMC5883L_INT || DOXYGEN */
+
+/**
+ * @brief   HMC5883L device initialization parameters
+ */
+typedef struct {
+
+    unsigned  dev;   /**< I2C device */
+    uint8_t   addr;  /**< I2C slave address */
+
+    gpio_t int_pin;  /**< DRDY interrupt pin: if #GPIO_UNDEF, interrupts are not used */
+
+    hmc5883l_meas_mode_t meas_mode; /**< Measurement mode (default #HMC5883L_MEAS_MODE_NORMAL) */
+    hmc5883l_meas_avg_t  meas_avg;  /**< Measurement avaraging (default #HMC5883L_MEAS_AVG_NONE) */
+    hmc5883l_dor_t       dor;       /**< Data output rate (default #HMC5883L_DOR_15) */
+    hmc5883l_op_mode_t   op_mode;   /**< Operation mode (#HMC5883L_OP_MODE_CONTINUOUS) */
+    hmc5883l_gain_t      gain;      /**< Gain (default #HMC5883L_GAIN_1090) */
+
+} hmc5883l_params_t;
+
+/**
+ * @brief   HMC5883L sensor device data structure type
+ */
+typedef struct {
+
+    hmc5883l_params_t params;    /**< device initialization parameters */
```

Comment:
If I see it correctly, there is actually no need in copying the complete params struct into RAM. Some of the static configuration values are never used again after init() was called, and others can be compacted...

Reply:
I will check.

Reply:
Yes, It is just the way as used in dozens of other drivers where the params are simply copied instead of picking single elements from them.

Reply:
Fixed, only params that are required later are copied to the device structure.